### DISCUSSION TITLE
Overlap between Freecad(v 1.0) and Blenderbim

### DISCUSSION THREAD

**Discussion by j209** [IfcOpenShell/IfcOpenShell#5347] [General]
The upcoming FreeCAD v1.0 and BlenderBIM both leverage IfcOpenShell for BIM workflows.
FreeCAD v1.0 is evolving into a dedicated, comprehensive BIM tool, fully integrating the NativeIFC standard for OpenBIM. It aims to provide a robust, modular, open-source solution for architects and engineers, with features supporting full lifecycle BIM processes.[NativeIFC 2024 Roadmap](https://github.com/yorikvanhavre/FreeCAD-NativeIFC/blob/main/doc/roadmap2024.md).

It is likely that there will be some overlap in functionality between the two applications.This can be confusing for both contributors and end users

**Comment by aothms** [comment]
Thanks for your message. I don't know if you have any immediate aims for your message. I'm personally not directly involved in the authoring apps.

One thing that could potentially be interesting is to consider IFC as a serialization format for (partial) exchange between the apps. We recently completed a GSOC project that examined adding a web-based interface to Bonsai (formerly known as BlenderBIM) using websockets. If we want to work more towards a complementary workflow that could be an entry point to have these apps talk with each other using standardized exchanges in a format that they anyway already support.

**Reply by j209** [reply]
why did the ifcopenshell team choose blender over freecad as the basis of ifcopenshell graphical interaction?

**Reply by duarteframos** [reply]
Pure speculation from my part but here are a few possible reasons:

Pros: 
- Blender has a comprehensive and versatile API
- Flexible and extensible UI
- High quality viewport and multiple shading modes
- Performant with high polygon counts and large amounts of objects
- Somewhat known in the industry and already used in AEC for archviz, rendering and viualization
- Good set of native tools and modelling facilities
- Actively developed
- Very customizable and configurable

Cons:
 - Not very good at precision modeling
 - No native solid based geometry
 - No good UI for 2D graphics editing or sheets management
 - No good UI for editing or viewing tabular or textual data

**Reply by Moult** [reply]
tl;dr Blender came first in the repo.

There wasn't really any "choosing", it just happened :) @aothms wrote a Blender import add-on back in May 2011 (!!!). FreeCAD later starting [using IfcOpenShell around Jan 2012](https://github.com/FreeCAD/FreeCAD/commit/4b51d8daa8c1fe4ced10f3f1a1304c502e9f5041). This was naturally developed in FreeCAD's repo. Then in late 2019 I started writing a Blender export add-on, also separately developed in my own repo. As the importer was not actively developed, I proposed to help maintain it and make the import / export capabilities align (originally, something you imported couldn't be exported). So that's why the code got merged into the IfcOpenShell repo and I deleted my own repo. As the importer and exporter grew closer, it then merged to form the first proof of concept of native IFC to form the BlenderBIM Add-on.

Since the beginning of the Native IFC approach, it was specifically designed so that anything we built would be agnostic of Blender where possible, so that we could then reuse the exact same engine in any other frontend: FreeCAD, Web, TKinter, etc. So overlap is a good thing, by design. There are many other tools out there as well, like IfcGref, IfcSuite...

**Reply by j209** [reply]
I think in order to have a better world(!!!:)) there should be a division work between these two apps and each of them should focus on the work in which they are better and have advantages (in the field of openbim).Freecad on designing,modeling, annotation ,creating 2d drawing from model ,create technical docs and bonsai on constrction simulation,clash detection,construction scheduling,estimating,structural analysis and...overal visualization.

**Reply by aothms** [reply]
@j209 I'm not sure I agree in full, there is a lot of value in experimentation. But I do agree that as foss developers we're building an ecosystem and not our own application silos. If you want to make your vision a reality I think you can directly contribute to figuring out these complementary workflows by taking data from one tool into the other, figuring out the things that break in the process so that they can get fixed, and showing the community that they can get the best of both worlds.

**Reply by duarteframos** [reply]
While I can see that developing for both Blender and FreeCAD in simultaneous may seem like wasted resources, having a mutually exclusive feature division may not prove advantageous in the long run if one can't complete its typical workflow with a minimum number of software.

Workflows that have to jump applications are always inefficient and error prone. Every time you have to switch from one software to another it wastes time, breaks mental flow, and something is always lost in translation.
Learning two new foreign looking software to accomplish a task that would otherwise be done in a single one (say a traditional 2D CAD program, or commercial BIM authoring applications, etc.) is not conducent to wider adoption.

If Bonsai and IFCOpenShell ever wants to find a modicum of success and be meaningfully recognized in the industry, users must find it appealing and easy to use. Part of that, I think, is eventually having a single comprehensive "nominal" package that the average non technical user can envision. This is important both in terms of marketing and practical use, spreading stuff across different software perceptually dilutes this ideal.

Imagine a conversation:

> User: Bah I'm so tired of paying subscriptions!
> You: Hey why don't you try open source alternative?
> U: I don't know how to.
> Y: Oh its easy, you just have to install Blender, and Bonsai, and FreeCAD, and Inkscape
> U: That sounds complicated.
> Y: ..and QCAD, and learn CSS, and script in Python, and...
> U: ...nevermind! *Quits*

This would discourage most potential users even if they were highly motivated. In other words, Bonsai and NativeIFC should each ideally appear as a comprehensive standalone solution to users problems, not another obstacle to overcome.

It will take a lot of time. Slowly but surely we will get there eventually
